src/Microsoft.ComponentDetection.Detectors/npm/NpmComponentDetectorWithRoots.cs

```diff
@@ -316,31 +340,114 @@ private void EnqueueDependencies(Queue<(JProperty Dependency, TypedComponent Par
+        TypedComponent parentComponent = null,
+        bool skipValidation = false)
     {
+        if (dependencies == null)
+        {
+            return true;
+        }
+
         var isValid = true;
-        if (dependencies != null)
+
+        foreach (var dependency in dependencies.Cast<JProperty>())
         {
-            foreach (var dependency in dependencies.Cast<JProperty>())
+            if (dependency?.Name == null)
             {
-                if (dependency == null || dependency.Name == null)
-                {
-                    continue;
-                }
+                continue;
+            }
 
-                var inLock = dependencyLookup.TryGetValue(dependency.Name, out var dependencyProperty);
-                if (inLock)
-                {
-                    queue.Enqueue((dependencyProperty, parentComponent));
-                }
-                else if (skipValidation)
+            var inLock = dependencyLookup.TryGetValue(lockFileVersion == 3 ? $"{NodeModules}/{dependency.Name}" : dependency.Name, out var dependencyProperty);
+            if (inLock)
+            {
+                queue.Enqueue((dependencyProperty, parentComponent));
+            }
+            else if (skipValidation)
+            {
+            }
+            else
+            {
+                isValid = false;
+            }
+        }
+
+        return isValid;
+    }
+
+    private bool TryEnqueueFirstLevelDependenciesLockfile3(
+        Queue<(JProperty DependencyProperty, TypedComponent ParentComponent)> queue,
+        JToken dependencies,
+        IDictionary<string, JProperty> dependencyLookup,
+        ISingleFileComponentRecorder componentRecorder,
+        TypedComponent parentComponent)
+    {
+        if (dependencies == null)
+        {
+            return true;
+        }
+
+        var isValid = true;
+
+        foreach (var dependency in dependencies.Cast<JProperty>())
+        {
+            if (dependency?.Name == null)
+            {
+                continue;
+            }
+
+            // First, check if there is an entry in the lockfile for this dependency nested in its ancestors
+            var ancestors = componentRecorder.DependencyGraph.GetAncestors(parentComponent.Id);
+            ancestors.Add(parentComponent.Id);
+
+            // remove version information
+            ancestors = ancestors.Select(x => x.Split(' ')[0]).ToList();
+
+            var possibleDepPaths = ancestors
+                .Select((t, i) => ancestors.TakeLast(ancestors.Count - i)); // depth-first search
+
+            var inLock = false;
+            JProperty dependencyProperty;
+            foreach (var possibleDepPath in possibleDepPaths)
+            {
+                var ancestorNodeModulesPath = string.Format(
+                    "{0}/{1}/{0}/{2}",
+                    NodeModules,
+                    string.Join($"/{NodeModules}/", possibleDepPath),
+                    dependency.Name);
+
```

Comment:
Is that second `NodeModules necessary? It looks like you would get `node_modules/node_modules/parent/node_modules/child`?

Reply:
Yes, its necessary. `string.Join` only places the string to be joined _between_ elements. So an array with 1 element doesn't get `node_modules` appended anywhere.